astro_reduction_tool.py

```diff
@@ -0,0 +1,222 @@
+            )
+            header_value = ccd_file.meta.get("combined")
+            if header_value is None or len(_ifc_mf) != 0:
+                data_list.append(ccd_file)
+                name_list.append(ccd_path)
+                now = datetime.now()
+                formatted_time = now.strftime("[%Y-%m-%dT%H:%M:%S.%f]")
+                logging.info(
+                    "%s  Loading %s file %s (1 of 1)",
+                    formatted_time,
+                    ccd_file.header["IMAGETYP"],
+                    ccd_path,
+                )
+    ccd_data_list = ccd_data(data_list,name_list)
+    return ccd_data_list
+
+def debias_dedark(file_list, master_dark, master_bias):
+    for h in range(0, len(file_list.data_list)):
+        ccd_file = file_list.data_list[h]
+        ccd_name = file_list.name_list[h]
+        bias_subtracted = ccdp.subtract_bias(ccd_file, master_bias)
+        file_list.data_list[h] = ccdp.subtract_dark(
+            bias_subtracted,
+            master_dark,
+            exposure_time="EXPTIME",
+            exposure_unit=u.adu,
+            scale=True,
+        )
+        now = datetime.now()
+        formatted_time = now.strftime("[%Y-%m-%dT%H:%M:%S.%f]")
+        logging.info(
+            "%s  Debias and dedark %s file %s (%d of %d)",
+            formatted_time,
+            ccd_file.header["IMAGETYP"],
+            ccd_name,
+            h+1,
+            len(file_list.data_list)
+        )
+    return file_list.data_list
+
+
+def master_flat_generator(flat_p, master_dark, master_bias, filt):
+    m_flat = []
+    m_flat_n = []
+    keys = ["imagetyp", "filter"]
+    for f in filt:
+        flat_data = read_path_into_CCDData(flat_p, keys, f)
+        flat_subtracted_list = debias_dedark(flat_data, master_dark, master_bias)
+        combined_flat = ccdp.combine(
+            flat_subtracted_list,
+            method="median",
+            unit=u.adu,
+            scale=inv_median,
+            sigma_clip=True,
+            sigma_clip_low_thresh=5.0,
+            sigma_clip_high_thresh=5.0,
+            sigma_clip_func=np.ma.median,
+            sigma_clip_dev_func=mad_std,
+            flat=2,
+        )
+        combined_flat.meta["combined"] = True
+        name = os.path.join(flat_p,"master_flat_{}.fits".format(f))
+        combined_flat.write(name, overwrite=True)
+        now = datetime.now()
+        formatted_time = now.strftime("[%Y-%m-%dT%H:%M:%S.%f]")
+        logging.info("%s  Created master_flat in %s filtr", formatted_time, f)
+        m_flat.append(combined_flat)
+        m_flat_n.append(name)
+        m_flat_list = ccd_data(m_flat,m_flat_n)
+    return m_flat_list
+
+
+def folders_reduction(dir_folder, m_dark, m_bias, m_flat_dictionary, d):
+    path = os.path.join(dir_folder,d)
+    ifc_folder = ccdp.ImageFileCollection(path, keywords=["imagetyp", "filter"])
+    filt = set(h["filter"] for h in ifc_folder.headers())
+    if os.path.exists(os.path.join(path, "pipeline_out")) is not True:
```

Comment:
1. Bardziej czytelnie byloby napisac:
```suggestion
    if not os.path.exists(os.path.join(path, "pipeline_out")):
```

Reply:
2. Zamiast w wielu miejscach stosowac `os.path.join(path, "pipeline_out")` warto przypisac wynik do zmiennej i uzywac jej dalej w kodzie

Reply:
3. Jesli uzyjesz `os.makedirs(..., exist_ok=True)` to nie musisz sprawdzac czy katalog juz istnieje.